suite/tests/client-interface/cleancall-opt-1.dll.c

```diff
@@ -0,0 +1,159 @@
+#else
+#define BINARY_NAME "client.cleancall-opt-1"
+#endif
+
+/* List of instrumentation functions. */
+#define FUNCTIONS() \
+        FUNCTION(modify_gprs) \
+        LAST_FUNCTION()
+
+#include "cleancall-opt-shared.h"
+
+static void event_exit(void);
+static dr_emit_flags_t event_basic_block(void *dc, void *tag, instrlist_t *bb,
+                                         bool for_trace, bool translating);
+
+DR_EXPORT void
+dr_init(client_id_t id)
+{
+    dr_register_exit_event(event_exit);
+    dr_register_bb_event(event_basic_block);
+    dr_fprintf(STDERR, "INIT\n");
+
+    /* Lookup pcs. */
+    lookup_pcs();
+    codegen_instrumentation_funcs();
+}
+
+static void
+event_exit(void)
+{
+    int i;
+    free_instrumentation_funcs();
+
+    for (i = 0; i < N_FUNCS; i++) {
+        DR_ASSERT_MSG(func_called[i],
+                      "Instrumentation function was not called!");
+    }
+    dr_fprintf(STDERR, "PASSED\n");
+}
+
+static dr_emit_flags_t
+event_basic_block(void *dc, void *tag, instrlist_t *bb,
+                  bool for_trace, bool translating)
+{
+    instr_t *entry = instrlist_first(bb);
+    app_pc entry_pc = instr_get_app_pc(entry);
+    int i;
+    bool inline_expected = false;
+    bool out_of_line_expected = false;
+    instr_t *before_label;
+    instr_t *after_label;
+
+    for (i = 0; i < N_FUNCS; i++) {
+        if (entry_pc == func_app_pcs[i])
+            break;
+    }
+    if (i == N_FUNCS)
+        return DR_EMIT_DEFAULT;
+
+    /* We're inserting a call to a function in this bb. */
+    func_called[i] = 1;
+    dr_insert_clean_call(dc, bb, entry, (void*)before_callee, false, 2,
+                         OPND_CREATE_INTPTR(func_ptrs[i]),
+                         OPND_CREATE_INTPTR(func_names[i]));
+
+    before_label = INSTR_CREATE_label(dc);
+    after_label = INSTR_CREATE_label(dc);
+
+    /* FIXME i#1569: passing instruction operands is NYI on AArch64.
+     * We use a workaround involving ADR. */
+    IF_AARCH64(save_current_pc(dc, bb, entry, &cleancall_start_pc, before_label));
+    PRE(bb, entry, before_label);
+    dr_insert_clean_call(dc, bb, entry, func_ptrs[i], false, 0);
+    PRE(bb, entry, after_label);
+    IF_AARCH64(save_current_pc(dc, bb, entry, &cleancall_end_pc, after_label));
+
+    dr_insert_clean_call(dc, bb, entry, (void*)after_callee, false, IF_X86_ELSE(6, 4),
+#ifdef X86
+                         opnd_create_instr(before_label),
+                         opnd_create_instr(after_label),
+#endif
+                         OPND_CREATE_INT32(inline_expected),
+                         OPND_CREATE_INT32(out_of_line_expected),
+                         OPND_CREATE_INT32(i),
+                         OPND_CREATE_INTPTR(func_names[i]));
+
+    return DR_EMIT_DEFAULT;
+}
+
+/*****************************************************************************/
+/* Instrumentation function code generation. */
+
+/*
+ * Modifies all GPRS and SIMD registers on X86.
+*/
```

Comment:
nit, comment style, could be in the same line.
If multi-line, start comment from the first line.
Also the */ alignment